src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+        assert isinstance(max_datetime, dt.datetime)
+        assert min_datetime <= max_datetime
+
         self.allow_naive = allow_naive
-        self.timezones = timezones
-        self.min_year = min_year or dt.MINYEAR
-        self.max_year = max_year or dt.MAXYEAR
-        for a in ['min_year', 'max_year']:
-            year = getattr(self, a)
-            if year < dt.MINYEAR:
-                raise InvalidArgument(u'%s out of range: %d < %d' % (
-                    a, year, dt.MINYEAR
-                ))
-            if year > dt.MAXYEAR:
-                raise InvalidArgument(u'%s out of range: %d > %d' % (
-                    a, year, dt.MAXYEAR
-                ))
+        self.timezones = tuple(timezones)
+        self.min_dt = min_datetime
+        self.max_dt = max_datetime
 
-    def do_draw(self, data):
+    @staticmethod
+    def _draw_naive(data, min_dt, max_dt):
+        # Draw a naive datetime between naive bounds
+        assert min_dt <= max_dt
         while True:
             try:
-                result = dt.datetime(
-                    year=cu.centered_integer_range(
-                        data, self.min_year, self.max_year, 2000
-                    ),
-                    month=cu.integer_range(data, 1, 12),
-                    day=cu.integer_range(data, 1, 31),
-                    hour=cu.integer_range(data, 0, 24),
-                    minute=cu.integer_range(data, 0, 59),
-                    second=cu.integer_range(data, 0, 59),
-                    microsecond=cu.integer_range(data, 0, 999999)
-                )
-                if (
-                    not self.allow_naive or
-                    (self.timezones and cu.boolean(data))
-                ):
-                    result = cu.choice(data, self.timezones).localize(result)
-                return result
-
-            except (OverflowError, ValueError):
-                pass
+                result = dict()
+                constrained_low = True
+                constrained_high = True
+                for name in ('year', 'month', 'day',
+                             'hour', 'minute', 'second', 'microsecond'):
+                    low = getattr(dt.datetime.min, name)
+                    high = getattr(dt.datetime.max, name)
+                    if constrained_low:
+                        low = max(low, getattr(min_dt, name))
+                    if constrained_high:
+                        high = min([high, getattr(max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result[name] = val
+                    constrained_low = constrained_low and val == low
+                    constrained_high = constrained_high and val == high
+                return dt.datetime(tzinfo=min_dt.tzinfo, **result)
+            except ValueError:
```

Comment:
Can we scope the `try … except` more tightly? I think we only expect it to be thrown on the `dt.datetime`.